Adding Pashto and Persian locales to psammead locales

ibMadbouly opened the pull request:
Resolves #1732 

**Overall change:** _A very high-level summary of easily-reproducible changes that can be understood by non-devs._
     

- This PR resolves the issue `Dual-calendar timestamps for Persian and Pashto `#1732 . 
- Extend the `momentjs`library to support `pashto` / `ps` locale as it is not supported natively, this locale supports the calendar the conversion from Gregorian timestamps to the corresponding `pashto` calendar in eastern numerals. 

- Update the existing `persian` / `fs ` locale to support the calendar conversion from Gregorian to
 its corresponding `persain` calendar with eastern numerals

**Code changes**

- At `psammead-locales/moment` added the following files

1. `ps.js` : extends the `moment` library to define new locale using `moment.defineLocale('ps',{/* ... */})` and perform the calendar converion from `gregorian` to `pashot`/`jalaali` using the library `jalaali-js` and the conversion is happening here `psammead-locales/moment/helpers/jalaali.js`, invoked from the `postFromat` function defined in the local

2. `fa.js`: update the existing locale in `moment js` to support the calendar conversion using `moment.updateLocale('fa',{/* ... */})` 

```js
postformat(string) {
     const str = jalaaliHelper.addJalaliDate(
      'ps',   // use `fa` for persain
      pashtoMonths,
      jalaliFromats,
      string,
    );

    if (!str.includes('datetime =')) {
      return stringHelper.useEasternNumerals(str);
    }
    return str;
  }
```

3. `ps.test.js`: adding test coverage for `ps.js` following the approach defined by `moment` for the existing locales.

4. `fa.test.js`: test coverage for the `fa.js`.

5. `/helpers/jalaali.js`: use the `jalaali-js` package to perform the calendar conversions

6. `/helpers/stringHelper.js`: added the numerals replacement logic

7.  `/helpers/stringHelper.test.js`: coverage for stringHelper

 

---

- [x] I have assigned myself to this PR and the corresponding issues
- [ ] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [ ] This PR requires manual testing

ibMadbouly committed:
adding pashto local to psammead-locales and adding tests for parsing

ibMadbouly committed:
add jalali formats and format test

ibMadbouly committed:
fixing numberSymbols and add ordinal tests

ibMadbouly committed:
adding format week testto pashto locale

ibMadbouly committed:
adding from test fro pashto locale at psammead locales

ibMadbouly committed:
adding tests

ibMadbouly committed:
adding tests to pashto psammead locale

dr3 commented on packages/utilities/psammead-locales/moment/ps.js:
```diff
@@ -0,0 +1,129 @@
+  months: 'جنوري_فبروري_مارچ_اپریل_می_جون_جولاې_اګست_سپتمبر_اکتوبر_نومبر_ډیسمبر'.split(
+    '_',
+  ),
+  monthsShort: 'جنوري_فبروري_مارچ_اپریل_می_جون_جولاې_اګست_سپتمبر_اکتوبر_نومبر_ډیسمبر'.split(
```
Hey :) What’s your source for these translations?

ibMadbouly replied:
Hey Drew :) , these translations are the same in the `moment-ws` repo here [https://github.com/bbc/moment-ws/blob/master/src/locale/ps.js](url)

Commit:
Merge branch 'latest' into adding-pashto-locale-to-psammead-locales

Commit:
Install jalaali-js dependancy

Signed-off-by: Rachael Njeri <rachael.njeri@andela.com>

Commit:
Fix linting issues

Signed-off-by: Rachael Njeri <rachael.njeri@andela.com>

ibMadbouly committed:
adding the persina  locale to psammead-locales + parse and format test

ibMadbouly committed:
adding additional test for  local

dr3 commented on packages/utilities/psammead-locales/moment/fa.js:
```diff
@@ -0,0 +1,52 @@
+  'اسفند',
+];
+
+const jalaliFromats = [
```
```suggestion
const jalaliFormats = [
```

ibMadbouly replied:
Thanks Drew

ibMadbouly committed:
fix typo

dr3 commented on packages/utilities/psammead-locales/moment/fa.test.js:
```diff
@@ -0,0 +1,193 @@
+/* eslint-disable */
```
Are there any tests in here that show the output for gregorian and jalali datestamps? I cant see them, but I dont speak farsi :P 🙈

ibMadbouly replied:
I applied some fixes, in the tests `test('format', function(){})` should show the calendar conversion for `ps` and `fa` . I don't speak farsi too 🙈

dr3 replied:
Sorry i still cant see them. Surely we should have tests thats like `moment(time).format('LL')` assert equals `<gregorian> <jalialli>`?

dr3 replied:
Also as far as I can see `test('format', function(){})` doesnt exist

jamesdonoh replied:
The test cases for dual calendar timestamps are here: https://github.com/bbc/psammead/pull/1844/files#diff-912d92b07d9011a084afbdaa2406e7c3R43-R45

ibMadbouly committed:
fixes in Pashto and Persian Locales

sareh commented on packages/utilities/psammead-locales/moment/fa.test.js:
```diff
@@ -0,0 +1,193 @@
+
+const assert = { equal: (val1, val2) => expect(val1).toEqual(val2) };
+
+test('parse', function () {
```
```suggestion
test('parse Gregorian months', function () {
```

sareh commented on packages/utilities/psammead-locales/moment/fa.test.js:
```diff
@@ -0,0 +1,193 @@
+    }
+});
+
+test('format', function () {
```
```suggestion
test('format dates and times', function () {
```

jamesdonoh replied:
This is for consistency with upstream.

sareh commented on packages/utilities/psammead-locales/moment/fa.test.js:
```diff
@@ -0,0 +1,193 @@
+    assert.equal(moment([2011, 0, 31]).format('DDDo'), '۳۱م', '۳۱');
+});
+
+test('format month', function () {
```
```suggestion
test('format Gregorian month', function () {
```

jamesdonoh replied:
This section looks verbatim from upstream https://github.com/moment/moment/blob/74f57907440cf9c4b119e6d5ad1f96b29acee1ce/src/test/locale/fa.js#L87-L92 so I suggest we preserve the naming.

sareh commented on packages/utilities/psammead-locales/moment/ps.js:
```diff
@@ -0,0 +1,127 @@
+  'dddd, D MMMM YYYY HH:mm', // LLLL FROMAT
+];
+moment.defineLocale('ps', {
+  months: 'جنوري_فبروري_مارچ_اپریل_می_جون_جولاې_اګست_سپتمبر_اکتوبر_نومبر_ډیسمبر'.split(
```
Might be worth adding a comment above this - that these are the Gregorian months.

jamesdonoh replied:
This was done.

sareh commented on packages/utilities/psammead-locales/moment/ps.js:
```diff
@@ -0,0 +1,127 @@
+  monthsShort: 'جنوري_فبروري_مارچ_اپریل_می_جون_جولاې_اګست_سپتمبر_اکتوبر_نومبر_ډیسمبر'.split(
+    '_',
+  ),
+  weekdays: 'اتوار_ ګول_نهه_شورو_زیارت_جمعه_خالي'.split('_'),
```
```suggestion
  weekdays: 'اتوار_ګول_نهه_شورو_زیارت_جمعه_خالي'.split('_'),
```

sareh commented on packages/utilities/psammead-locales/moment/ps.js:
```diff
@@ -0,0 +1,127 @@
+    '_',
+  ),
+  weekdays: 'اتوار_ ګول_نهه_شورو_زیارت_جمعه_خالي'.split('_'),
+  weekdaysShort: 'اتوار_ ګول_نهه_شورو_زیارت_جمعه_خالي '.split('_'),
```
```suggestion
  weekdaysShort: 'اتوار_ګول_نهه_شورو_زیارت_جمعه_خالي '.split('_'),
```

sareh commented on packages/utilities/psammead-locales/moment/ps.test.js:
```diff
@@ -0,0 +1,234 @@
+moment.locale('ps');
+const assert = { equal: (val1, val2) => expect(val1).toEqual(val2) };
+
+test('parse', function () {
```
```suggestion
test('parse Gregorian months', function () {
```

sareh commented on packages/utilities/psammead-locales/moment/ps.test.js:
```diff
@@ -0,0 +1,234 @@
+    assert.equal(moment([2011, 0, 31]).format('DDDo'), '۳۱', '۳۱');
+});
+
+test('format month', function () {
```
```suggestion
test('format Gregorian month', function () {
```

sareh commented on packages/utilities/psammead-locales/moment/helpers/jalaali.js:
```diff
@@ -0,0 +1,30 @@
+  return output;
+}
+
+function addJalaliDate(locale, jalaliMonths, jalaliFormats, gregorianString) {
```
This function will not be needed in this format.
The reason is, the markup for the two dates on a page should be this, with the gregorian calendar in a time element with the datetime that search engines use to get the date value. Then a dash with the Jalaali date in another time element. 
```
<time datetime="2019-01-31">۱ ژانویه ۲۰۱۹</time> - <time>۱۱ بهمن ۱۳۹۷</time>
```

sareh replied:
To clarify, we only want these utilities to export the dates as strings, not as components with `<time>` tags. It will be within the `TimestampContainer` package where we use the `<time>` element. 


We'll want the API to be something that is simple to use, e.g.:

`jalaaliPersianDatetime(unixTimestamp)`
`jalaaliPashtoDatetime(unixTimestamp)`
And these exports can abstract away the usage of the `stringHelper.useEasternNumerals`, `getJalaaliDatetime` and `jalaliMonths`

jamesdonoh replied:
Agree we don't want to include HTML elements here, though the API can't be exactly as suggested here as the implementation (as in `moment-ws`) relies on Moment.js' `postformat` hook to append the extra information after the timestamp has been formatted.

jamesdonoh replied:
Discussed this with @sareh and we decided it's OK to return a single human-readable string to be the content of a a single `time` element. Therefore we do not require an API to return both timestamps separately.

sareh commented on packages/utilities/psammead-locales/moment/helpers/jalaali.js:
```diff
@@ -0,0 +1,30 @@
+const moment = require('moment-timezone/moment-timezone');
+const jalali = require('jalaali-js');
+
+function getJalaliString(gregorianMoment, jalaliMonths) {
```
```suggestion
function getJalaaliDatetime(gregorianMoment, jalaliMonths) {
```
Can you please export this function directly. 
It will be good to add a few unit tests for this. For both fa and ps locales using this Jalaali helper. 
E.g. With an input date of 31-01-2019, should get for fa locale:

<span dir="rtl">۱۱ بهمن ۱۳۹۷</span>
And should get, for ps locale:
<span dir="rtl">۱۱ سلواغه ۱۳۹۷</span>

jamesdonoh replied:
Do you think the existing tests such as https://github.com/bbc/psammead/pull/1844/files#diff-96d07ccc729cfbdc689e14884f8452c9R47-R60 provide sufficient coverage for the helper itself?

I would agree the tests above could be rewritten to be more generic as the helper is actually agnostic to the locale and format, but specific coverage for Persian and Pashto is provided by `fa.test.js` and `pa.test.js`.

sareh reviewed (changes requested):
Looks great so far!
I've added some suggestions for unit tests for the Jalaali helper. Also, I've updated the test descriptions so that people who can't read Persian or Pashto can know which variant is being tested there.

ibMadbouly committed:
adding more clarifications for the tests and locales

ibMadbouly committed:
fix pashto test

ibMadbouly committed:
fix persian test

ibMadbouly committed:
adding tests for getgetJalaaliDatetime and change name

ibMadbouly committed:
adding L format for pashto and persian locales

ibMadbouly committed:
adding stringHelper and its test and use it in jalaali testing

ibMadbouly committed:
add the support of time elements for calendar conversion

ibMadbouly committed:
make use of stringHelper

Commit:
Merge branch 'latest' into adding-pashto-locale-to-psammead-locales

jamesdonoh committed:
Update packages/utilities/psammead-locales/moment/ps.js

CR suggestion from @sareh

Co-Authored-By: Sareh <sareh@users.noreply.github.com>

jamesdonoh committed:
Merge branch 'latest' into adding-pashto-locale-to-psammead-locales

jamesdonoh committed:
Merge branch 'latest' into adding-pashto-locale-to-psammead-locales

jamesdonoh committed:
Merge branch 'latest' into adding-pashto-locale-to-psammead-locales

dr3 commented on packages/utilities/psammead-locales/moment/fa.test.js:
```diff
@@ -0,0 +1,228 @@
+
+const assert = { equal: (val1, val2) => expect(val1).toEqual(val2) };
+/*
+test('parse Gregorian months', function () {
```
Why is this commented out?

jamesdonoh replied:
Now uncommented.

jamesdonoh committed:
Update test to match change in 50be143

jamesdonoh committed:
Uncomment tests

jamesdonoh committed:
Remove <time> elements from output

We do not want to add these elements in this package, this will be taken care of
in TimestampContainer

jamesdonoh committed:
Only include dual calendar for formats with long month name

jamesdonoh committed:
Merge branch 'latest' of github.com:bbc/psammead into adding-pashto-locale-to-psammead-locales

jamesdonoh committed:
Merge branch 'latest' of github.com:bbc/psammead into adding-pashto-locale-to-psammead-locales

jamesdonoh committed:
Style/formatting updates for consistency

jamesdonoh committed:
Merge branch 'latest' of github.com:bbc/psammead into adding-pashto-locale-to-psammead-locales

jamesdonoh committed:
Use Eastern numerals unconditionally

jamesdonoh committed:
Merge branch 'latest' of github.com:bbc/psammead into adding-pashto-locale-to-psammead-locales

jamesdonoh committed:
Unintended changes from merge

jamesdonoh committed:
Amend numeral 4 to other version

jamesdonoh committed:
Reset tests to be closer to upstream for easier comparison

jamesdonoh committed:
Merge branch 'latest' of github.com:bbc/psammead into adding-pashto-locale-to-psammead-locales

jamesdonoh committed:
Put Jalaali first to make handling time components easier

jamesdonoh committed:
Limit Jalaali date to be only added to date formats not times

This is because the current method of re-parsing the generated string
to see if it's valid does not work if there is a time component.

jamesdonoh committed:
Remove unverified parts of Pashto locale/tests

jamesdonoh committed:
Merge branch 'latest' of github.com:bbc/psammead into adding-pashto-locale-to-psammead-locales

jamesdonoh committed:
Reset to latest

dr3 commented on packages/utilities/psammead-locales/moment/ps.test.js:
```diff
@@ -0,0 +1,22 @@
+
+test('format', function () {
+    var a = [
+            ['LL',                                 '۲۵ سلواغه ۱۳۸۸ - فبروري ۱۴، ۲۰۱۰'],
```
Could you maybe add a couple other formats to ensure it doesnt attach the dual data to like `MMM` for example?

jamesdonoh replied:
Good idea, added some negative cases. Although note this is not currently a fully-functioning moment locale.

dr3 commented on packages/utilities/psammead-locales/package.json:
```diff
@@ -25,6 +25,10 @@
   ],
+  "dependencies": {
+    "jalaali-js": "1.0.0",
+    "moment-timezone": "^0.5.26"
```
I think `moment-timezone` should be a peer dep

jamesdonoh replied:
Amended.

dr3 commented on packages/utilities/psammead-locales/moment/helpers/jalaali.test.js:
```diff
@@ -0,0 +1,75 @@
+});
+
+test('getJalaaliDatetime for persian locale', function() {
+  const gregorinaMonent = moment(
```
`gregorinaMonent` typo

jamesdonoh replied:
Fixed.

dr3 reviewed (changes requested):
LGTM, just a couple tweaks. Im gunna miss something simply by not being able to speak these languages, but as far as i can tell, works great :D Awesome work

RayNjeri commented:
I think this branch is behind latest, and needs to be updated, also do not forget to bump @psammead-locales and update the Changelog.md @jamesdonoh

jamesdonoh committed:
Version bump and documentation update

jamesdonoh committed:
Merge branch 'latest' of github.com:bbc/psammead into adding-pashto-locale-to-psammead-locales

jamesdonoh commented:
OK, thanks for reminder @RayNjeri

jamesdonoh committed:
Add negative tests cases for Jalaali calendar

jamesdonoh committed:
Fix typo

jamesdonoh committed:
Make moment+timezone peerDeps

jamesdonoh committed:
Merge branch 'latest' of github.com:bbc/psammead into adding-pashto-locale-to-psammead-locales

jamesdonoh commented:
> Im gunna miss something simply by not being able to speak these languages

Me too, hoping @sareh can offer some certainty around https://github.com/bbc/psammead/pull/1844/files#diff-912d92b07d9011a084afbdaa2406e7c3R29-R52 otherwise we can get Editorial to check once it's on a URL.

dr3 commented on packages/utilities/psammead-locales/README.md:
```diff
@@ -4,9 +4,22 @@ This package provides a collection of locale configs, used in BBC World Service
+- `/moment/ar` - locale override for Arabic
+- `/moment/fa` - locale override for Persian (Farsi)
+- `/moment/yo` - locale override for Yoruba
+- `/moment/ps` - locale override for Pashto
```
Pashto isnt an override, but a new locale right?

jamesdonoh replied:
Good spot, fixed.

jamesdonoh committed:
Reclassify Pashto locale

dr3 reviewed (approved):
![](https://media.giphy.com/media/XbxZ41fWLeRECPsGIJ/giphy.gif)

sareh commented on packages/utilities/psammead-locales/moment/fa.test.js:
```diff
@@ -0,0 +1,214 @@
+    assert.equal(start.from(moment([2007, 1, 28]).add({d: 548}), true), '۲ سال',       '548 days = 2 years');
+    assert.equal(start.from(moment([2007, 1, 28]).add({y: 1}), true),   'یک سال',      '1 year = a year');
+    assert.equal(start.from(moment([2007, 1, 28]).add({y: 5}), true),   '۵ سال',       '5 years = 5 years');
+});
```
All translations for `fa` locale read correctly.

sareh reviewed (approved):
👍 Great work with simplifying the number of additions made to the locales. The `fa` locale translations in the tests read correctly.

sareh committed:
Merge branch 'latest' into adding-pashto-locale-to-psammead-locales

sareh commented on packages/utilities/psammead-locales/moment/helpers/jalaali.js:
```diff
@@ -0,0 +1,31 @@
+const moment = require('moment-timezone/moment-timezone');
```
Does it not need to be this?
```suggestion
const moment = require('moment-timezone');
```

jamesdonoh replied:
This suggested change didn't fix the CI errors for me as `moment-timezone` was not available. I looked into it and I'm not sure why `moment-timezone` is even referenced, the line originates from https://github.com/bbc/moment-ws/pull/21 but none of the timezone API is used here AFAICS.

So instead in 265362b19ba8efce4c09a916b34facd41a76fc3d I've made `moment` a full dependency (since this package explicitly requires it) and removed `moment-timezone`. This seems correct to me but let me know if I'm missing something or you can see a better way. /cc @dr3

sareh commented on packages/utilities/psammead-locales/moment/helpers/jalaali.test.js:
```diff
@@ -0,0 +1,75 @@
+/* eslint-disable func-names */
+const moment = require('moment-timezone/moment-timezone');
```
```suggestion
const moment = require('moment-timezone');
```

jamesdonoh committed:
Declare moment as dependency for Jalaali

moment must be explicitly required for the Jalaali helper to work. moment-timezone does not seem to actually be need and is a remnant from moment-ws

jamesdonoh committed:
Merge branch 'adding-pashto-locale-to-psammead-locales' of github.com:bbc/psammead into adding-pashto-locale-to-psammead-locales

jamesdonoh committed:
Merge branch 'latest' of github.com:bbc/psammead into adding-pashto-locale-to-psammead-locales

staylos92 committed:
Merge branch 'latest' into adding-pashto-locale-to-psammead-locales
